src/main/java/bio/terra/service/dataset/flight/update/DatasetSchemaUpdateAddColumnsBigQueryStep.java

```diff
@@ -0,0 +1,96 @@
+
+import bio.terra.common.Column;
+import bio.terra.model.ColumnModel;
+import bio.terra.model.DatasetSchemaUpdateModel;
+import bio.terra.service.dataset.Dataset;
+import bio.terra.service.dataset.DatasetDao;
+import bio.terra.service.dataset.DatasetJsonConversion;
+import bio.terra.service.dataset.DatasetTable;
+import bio.terra.service.tabulardata.google.BigQueryProject;
+import bio.terra.service.tabulardata.google.bigquery.BigQueryDatasetPdao;
+import bio.terra.service.tabulardata.google.bigquery.BigQueryPdao;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import com.google.cloud.bigquery.BigQuery;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+public class DatasetSchemaUpdateAddColumnsBigQueryStep implements Step {
+  private final BigQueryDatasetPdao bigQueryDatasetPdao;
+  private final DatasetDao datasetDao;
+  private final UUID datasetId;
+  private final DatasetSchemaUpdateModel updateModel;
+
+  private static final List<String> EMPTY_LIST = Collections.emptyList();
+
+  public DatasetSchemaUpdateAddColumnsBigQueryStep(
+      BigQueryDatasetPdao bigQueryDatasetPdao,
+      DatasetDao datasetDao,
+      UUID datasetId,
+      DatasetSchemaUpdateModel updateModel) {
+    this.bigQueryDatasetPdao = bigQueryDatasetPdao;
+    this.datasetDao = datasetDao;
+    this.datasetId = datasetId;
+    this.updateModel = updateModel;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    Dataset dataset = datasetDao.retrieve(datasetId);
+
+    BigQueryProject bigQueryProject = BigQueryProject.from(dataset);
+    BigQuery bigQuery = bigQueryProject.getBigQuery();
+    String datasetName = BigQueryPdao.prefixName(dataset.getName());
+
+    for (var columnChanges : updateModel.getChanges().getAddColumns()) {
+      DatasetTable table = dataset.getTableByName(columnChanges.getTableName()).orElseThrow();
+      for (ColumnModel columnModel : columnChanges.getColumns()) {
+        Column column = DatasetJsonConversion.columnModelToDatasetColumn(columnModel, EMPTY_LIST);
+        bigQueryDatasetPdao.createColumn(datasetName, bigQuery, table, column);
+      }
+      bigQuery.update(
+          BigQueryDatasetPdao.buildLiveView(bigQueryProject.getProjectId(), datasetName, table));
+    }
+    return StepResult.getStepResultSuccess();
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    Dataset dataset = datasetDao.retrieve(datasetId);
+
+    BigQueryProject bigQueryProject = BigQueryProject.from(dataset);
+    BigQuery bigQuery = bigQueryProject.getBigQuery();
+    String datasetName = BigQueryPdao.prefixName(dataset.getName());
+
+    for (var columnChanges : updateModel.getChanges().getAddColumns()) {
```

Comment:
Can the same table be referenced by more than one column change? If so, would you need to revert these operations in reverse order, if one change depends on a previous one?

Reply:
Can you give an example of a case where two column additions to the same table depend on each other? 🤔